zookeeper-api/src/main/java/org/apache/helix/zookeeper/util/HttpRoutingDataReader.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.codehaus.jackson.map.ObjectMapper;
+
+
+public class HttpRoutingDataReader {
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private static volatile Map<String, List<String>> _rawRoutingData;
+  private static volatile MetadataStoreRoutingData _metadataStoreRoutingData;
+
+  /**
+   * This class is a Singleton.
+   */
+  private HttpRoutingDataReader() {
+  }
+
+  /**
+   * Fetches routing data from the data source via HTTP.
+   * @return a mapping from "metadata store realm addresses" to lists of
+   * "metadata store sharding keys", where the sharding keys in a value list all route to
+   * the realm address in the key disallows a meaningful mapping to be returned
+   */
+  @RetryOnFailure(attempts = 3, delay = 1, unit = TimeUnit.SECONDS)
+  public static Map<String, List<String>> getRawRoutingData()
+      throws IOException {
+    if (_rawRoutingData == null) {
+      synchronized (HttpRoutingDataReader.class) {
+        if (_rawRoutingData == null) {
+          String msdsEndpoint =
+              System.getProperty(MetadataStoreRoutingConstants.MSDS_SERVER_ENDPOINT_KEY);
+          if (msdsEndpoint == null || msdsEndpoint.isEmpty()) {
+            throw new IllegalStateException(
+                "HttpRoutingDataReader was unable to find a valid MSDS endpoint String in System Properties!");
+          }
+
+          // Note: HttpClient's timeout settings are system timeout settings by default
+          CloseableHttpClient httpClient = HttpClients.createDefault();
+          // Update the reference
+          _rawRoutingData = getAllRoutingData(httpClient, msdsEndpoint);
+          // Close any open resources
+          httpClient.close();
+        }
+      }
+    }
+    return _rawRoutingData;
+  }
+
+  /**
+   * Returns the routing data read from MSDS in a MetadataStoreRoutingData format.
+   * @return
+   * @throws IOException
+   * @throws InvalidRoutingDataException
+   */
+  public static MetadataStoreRoutingData getMetadataStoreRoutingData()
+      throws IOException, InvalidRoutingDataException {
+    if (_metadataStoreRoutingData == null) {
+      synchronized (HttpRoutingDataReader.class) {
+        if (_metadataStoreRoutingData == null) {
+          _metadataStoreRoutingData = new TrieRoutingData(getRawRoutingData());
+        }
+      }
+    }
+    return _metadataStoreRoutingData;
+  }
+
+  /**
+   * Makes an HTTP call to fetch all routing data.
+   * @param httpClient
+   * @param msdsEndpoint
+   * @return
+   * @throws IOException
+   */
+  private static Map<String, List<String>> getAllRoutingData(CloseableHttpClient httpClient,
+      String msdsEndpoint)
+      throws IOException {
+    HttpGet requestAllData = new HttpGet(
+        msdsEndpoint + MetadataStoreRoutingConstants.MSDS_GET_ALL_ROUTING_DATA_ENDPOINT);
+    CloseableHttpResponse response = httpClient.execute(requestAllData);
```

Comment:
will httpCient.execute throw other exceptions aside from IOException?

Reply:
The other more important thing is that I assume that the user needs to configure multiple host to msdsEndPoint? Similar like DNS return IP address set?
Otherwise, this is a single point of failure. 
How doe it work currently in rest client? We can leverage the same mechanism to avoid SPOF.

Reply:
I don't think so.

Reply:
It's a VIP so we have different hosts that are mapped to the single endpoint. But this is an implementation detail for Helix applications/enterprises, so let us now discuss those details here.

Reply:
Either I did not get what you mean by VIP or you did not get what I mean here. What really concerns me is that some rest servers may not be available for whatever reason, say network partition. Then we should retry for another rest server. Otherwise, mission critical application may not be able to start. 

So the retry logic is very critical. We don't have this retry logic. You can think of this retry logic similar to Zookeeper native client retry different quorum node.

Reply:
Discussed offline. VIP is just an example - different network vendors will provide different abstraction layers.